FileSystemSyncAccessHandle::{read,write}'s option should be optional

Constellation opened the pull request:
#### 08343037ad9b7c9a6c75d6c51a791755a7522e4c
<pre>
FileSystemSyncAccessHandle::{read,write}&apos;s option should be optional
<a href="https://bugs.webkit.org/show_bug.cgi?id=249631">https://bugs.webkit.org/show_bug.cgi?id=249631</a>
rdar://103541719

Reviewed by Justin Michaud.

The current FileSystemSyncAccessHandle::{read,write} implementation has a bug that `options` parameter is specified as mandatory.
As a result, these APIs throw errors because it is using `read` / `write` functions without options.
This patch fixes the implementation to make them optional.

When optional `at` is not found, then the behavior is using the current file position&apos;s cursor. Thus, we just skip seekFile if
it is not specified.

* LayoutTests/storage/filesystemaccess/resources/sync-access-handle-read-write.js:
(write):
(read):
(async test):
* LayoutTests/storage/filesystemaccess/sync-access-handle-read-write-worker-expected.txt:
* Source/WebCore/Modules/filesystemaccess/FileSystemSyncAccessHandle.cpp:
(WebCore::FileSystemSyncAccessHandle::read):
(WebCore::FileSystemSyncAccessHandle::write):
* Source/WebCore/Modules/filesystemaccess/FileSystemSyncAccessHandle.h:
* Source/WebCore/Modules/filesystemaccess/FileSystemSyncAccessHandle.idl:

Canonical link: <a href="https://commits.webkit.org/258160@main">https://commits.webkit.org/258160@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e90371a64faf49e8a9073958dd02c6cc71d9e33d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101040 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10196 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34095 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110343 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/170600 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/105027 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11126 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/1070 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/93478 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108177 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106823 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8428 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/91690 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/35043 "layout-tests (failure)") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/90333 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/23086 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/78033 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3864 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24603 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3892 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/1007 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10004 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44111 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5660 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2940 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud reviewed (approved):
r=me

szewai commented on Source/WebCore/Modules/filesystemaccess/FileSystemSyncAccessHandle.h:
```diff
@@ -43,7 +43,7 @@ template<typename> class DOMPromiseDeferred;
 public:
     struct FilesystemReadWriteOptions {
-        unsigned long long at;
+        std::optional<unsigned long long> at;
```
It seems we should make `read()` and `write()` take `std::optional<FilesystemReadWriteOptions>` instead of making `at` optional. We could probably make `at` initialized to 0.

Constellation replied:
@szewai are we supporting `std::optional<>` with dictionary? Existing code is not using it for `optional Dictionary` case.
And if we always initialize it to zero & if we cannot get `std::optional<Dictionary>`, then we cannot distinguish between non-passed argument v.s. explicitly-passed `0`.

szewai replied:
hmm I think `std::optional` is used with dictionary in FileSystemDirectoryHandle.idl and Navigator+CookieConsent.idl. Is the usage unexpected?

Constellation replied:
Yes, I think `FileSystemDirectoryHandle.idl`'s code is wrong since the generated code is,
```
  317 static inline JSC::EncodedJSValue jsFileSystemDirectoryHandlePrototypeFunction_removeEntryBody(JSC::JSGlobalObject* lexicalGlobalObject, JSC::CallFrame* callFrame, typename IDLOperationRet      urningPromise<JSFileSystemDirectoryHandle>::ClassParameter castedThis, Ref<DeferredPromise>&& promise)
  318 {
  319     auto& vm = JSC::getVM(lexicalGlobalObject);
  320     auto throwScope = DECLARE_THROW_SCOPE(vm);
  321     UNUSED_PARAM(throwScope);
  322     UNUSED_PARAM(callFrame);
  323     auto& impl = castedThis->wrapped();
  324     if (UNLIKELY(callFrame->argumentCount() < 1))
  325         return throwVMError(lexicalGlobalObject, throwScope, createNotEnoughArgumentsError(lexicalGlobalObject));
  326     EnsureStillAliveScope argument0 = callFrame->uncheckedArgument(0);
  327     auto name = convert<IDLUSVString>(*lexicalGlobalObject, argument0.value());
  328     RETURN_IF_EXCEPTION(throwScope, encodedJSValue());
  329     EnsureStillAliveScope argument1 = callFrame->argument(1);
  330     auto options = convert<IDLDictionary<FileSystemDirectoryHandle::RemoveOptions>>(*lexicalGlobalObject, argument1.value());
  331     RETURN_IF_EXCEPTION(throwScope, encodedJSValue());
  332     RELEASE_AND_RETURN(throwScope, JSValue::encode(toJS<IDLPromise<IDLUndefined>>(*lexicalGlobalObject, *castedThis->globalObject(), throwScope, [&]() -> decltype(auto) { return impl.remov      eEntry(WTFMove(name), WTFMove(options), WTFMove(promise)); })));
  333 }
```
```
  116 template<> FileSystemDirectoryHandle::RemoveOptions convertDictionary<FileSystemDirectoryHandle::RemoveOptions>(JSGlobalObject& lexicalGlobalObject, JSValue value)
  117 {
  118     VM& vm = JSC::getVM(&lexicalGlobalObject);
  119     auto throwScope = DECLARE_THROW_SCOPE(vm);
  120     bool isNullOrUndefined = value.isUndefinedOrNull();
  121     auto* object = isNullOrUndefined ? nullptr : value.getObject();
  122     if (UNLIKELY(!isNullOrUndefined && !object)) {
  123         throwTypeError(&lexicalGlobalObject, throwScope);
  124         return { };
  125     }
  126     FileSystemDirectoryHandle::RemoveOptions result;
  127     JSValue recursiveValue;
  128     if (isNullOrUndefined)
  129         recursiveValue = jsUndefined();
  130     else {
  131         recursiveValue = object->get(&lexicalGlobalObject, Identifier::fromString(vm, "recursive"_s));
  132         RETURN_IF_EXCEPTION(throwScope, { });
  133     }
  134     if (!recursiveValue.isUndefined()) {
  135         result.recursive = convert<IDLBoolean>(lexicalGlobalObject, recursiveValue);
  136         RETURN_IF_EXCEPTION(throwScope, { });
  137     } else
  138         result.recursive = false;
  139     return result;
  140 }
  ```
So, `auto options = convert<IDLDictionary<FileSystemDirectoryHandle::RemoveOptions>>` is not propagating optional information. And it is always creating a dictionary.

szewai replied:
ah I should correct that instead; thanks for pointing out

szewai commented on LayoutTests/storage/filesystemaccess/resources/sync-access-handle-read-write.js:
```diff
@@ -45,41 +51,84 @@ async function test()
-        read(accessHandle, totalReadSize, totalWriteSize - totalReadSize, "This is second sentence.");
-        read(accessHandle, 0, totalWriteSize, "This is first sentence.This is second sentence.");
-
-        // Wrong offset to read and write.
```
It seems we could insert a test case here without duplicating the whole test.
Since file cursor should be updated after `read()`, we could add `write()` without options here, and verify it appends to the end of the file by `read()`.
Something like:
```
totalReadSize = totalWriteSize;
// File cursor should be updated by last read(), so the next write() should append to file.
totalWriteSize += write(accessHandle, null, "This is third sentence.");
read(accessHandle, 0, totalReadSize, "This is first sentence.This is second sentence.");
read(accessHandle, null, totalWriteSize - totalReadSize, "This is third sentence.");
```

Constellation replied:
Nice, changed.

Constellation committed:
FileSystemSyncAccessHandle::{read,write}'s option should be optional
https://bugs.webkit.org/show_bug.cgi?id=249631
rdar://103541719

Reviewed by Justin Michaud.

The current FileSystemSyncAccessHandle::{read,write} implementation has a bug that `options` parameter is specified as mandatory.
As a result, these APIs throw errors because it is using `read` / `write` functions without options.
This patch fixes the implementation to make them optional.

When optional `at` is not found, then the behavior is using the current file position's cursor. Thus, we just skip seekFile if
it is not specified.

* LayoutTests/storage/filesystemaccess/resources/sync-access-handle-read-write.js:
(write):
(read):
(async test):
* LayoutTests/storage/filesystemaccess/sync-access-handle-read-write-worker-expected.txt:
* Source/WebCore/Modules/filesystemaccess/FileSystemSyncAccessHandle.cpp:
(WebCore::FileSystemSyncAccessHandle::read):
(WebCore::FileSystemSyncAccessHandle::write):
* Source/WebCore/Modules/filesystemaccess/FileSystemSyncAccessHandle.h:
* Source/WebCore/Modules/filesystemaccess/FileSystemSyncAccessHandle.idl:

Canonical link: https://commits.webkit.org/258160@main
